crates/chain/src/indexed_tx_graph.rs

```diff
@@ -252,61 +252,62 @@ where
     }
 }
 
-/// Methods are available if the anchor (`A`) implements [`AnchorFromBlockPosition`].
-impl<A: Anchor, I: Indexer> IndexedTxGraph<A, I>
+/// Methods are available if the anchor (`A`) can be created from [`BlockPosition`].
+impl<'b, A, I> IndexedTxGraph<A, I>
 where
     I::ChangeSet: Default + Merge,
-    A: AnchorFromBlockPosition,
+    A: Anchor + From<BlockPosition<'b>>,
+    I: Indexer,
 {
     /// Batch insert all transactions of the given `block` of `height`, filtering out those that are
     /// irrelevant.
     ///
-    /// Each inserted transaction's anchor will be constructed from
-    /// [`AnchorFromBlockPosition::from_block_position`].
+    /// Each inserted transaction's anchor will be constructed using [`BlockPosition`].
     ///
     /// Relevancy is determined by the internal [`Indexer::is_tx_relevant`] implementation of `I`.
     /// Irrelevant transactions in `txs` will be ignored.
     pub fn apply_block_relevant(
-        &mut self,
-        block: &Block,
+        &'b mut self,
+        block: &'b Block,
         height: u32,
     ) -> ChangeSet<A, I::ChangeSet> {
         let block_id = BlockId {
             hash: block.block_hash(),
             height,
         };
         let mut changeset = ChangeSet::<A, I::ChangeSet>::default();
-        for (tx_pos, tx) in block.txdata.iter().enumerate() {
+        for tx in &block.txdata {
             changeset.indexer.merge(self.index.index_tx(tx));
             if self.index.is_tx_relevant(tx) {
                 let txid = tx.compute_txid();
-                let anchor = A::from_block_position(block, block_id, tx_pos);
                 changeset.tx_graph.merge(self.graph.insert_tx(tx.clone()));
-                changeset
-                    .tx_graph
-                    .merge(self.graph.insert_anchor(txid, anchor));
+                changeset.tx_graph.merge(
+                    self.graph
+                        .insert_anchor(txid, BlockPosition::new(block, block_id).into()),
+                );
             }
         }
         changeset
     }
 
     /// Batch insert all transactions of the given `block` of `height`.
     ///
-    /// Each inserted transaction's anchor will be constructed from
-    /// [`AnchorFromBlockPosition::from_block_position`].
+    /// Each inserted transaction's anchor will be constructed using [`BlockPosition`].
     ///
     /// To only insert relevant transactions, use [`apply_block_relevant`] instead.
     ///
     /// [`apply_block_relevant`]: IndexedTxGraph::apply_block_relevant
-    pub fn apply_block(&mut self, block: Block, height: u32) -> ChangeSet<A, I::ChangeSet> {
+    pub fn apply_block(&'b mut self, block: &'b Block, height: u32) -> ChangeSet<A, I::ChangeSet> {
```

Comment:
This can take a `BlockPosition` (what I would call `BlockAndHeight`).

Reply:
@LLFourn We shouldn't do this if `BlockPosition` (which I'm thinking should be called `TxPosInBlock`) needs `tx_pos` field.